src/firecracker/src/api_server/request/hotplug.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+use vmm::logger::{IncMetric, METRICS};
+use vmm::rpc_interface::VmmAction;
+use vmm::vmm_config::hotplug::{HotplugVcpuConfig, HotplugVcpuError};
+
+use super::super::parsed_request::{ParsedRequest, RequestError};
+use super::{Body, Method, StatusCode};
+
+pub(crate) fn parse_put_hotplug(
+    body: &Body,
+    resource_type_from_path: Option<&str>,
+) -> Result<ParsedRequest, RequestError> {
+    METRICS.put_api_requests.hotplug.inc();
+    match resource_type_from_path {
+        // Memory hotplugging not implemented yet.
+        Some(resource) => match resource {
+            "vcpu" => parse_put_vcpu_hotplug(body),
+            _ => Err(RequestError::InvalidPathMethod(
+                format!("/hotplug/{}", resource),
+                Method::Put,
+            )),
+        },
+        None => Err(RequestError::Generic(
+            StatusCode::BadRequest,
+            "Missing resource type to hotplug.".to_string(),
+        )),
+    }
+}
+
+fn parse_put_vcpu_hotplug(body: &Body) -> Result<ParsedRequest, RequestError> {
+    METRICS.hotplug.vcpu_hotplug_request_count.inc();
+    let config = serde_json::from_slice::<HotplugVcpuConfig>(body.raw()).map_err(|err| {
+        METRICS.hotplug.vcpu_hotplug_request_fails.inc();
+        err
+    })?;
+
+    Ok(ParsedRequest::new_sync(VmmAction::HotplugVcpu(config)))
+}
+
+#[cfg(test)]
+mod tests {
+
+    use hotplug::{parse_put_hotplug, parse_put_vcpu_hotplug};
+    use vmm::{logger::{IncMetric, METRICS}, rpc_interface::VmmAction, vmm_config::hotplug::{HotplugVcpuConfig, HotplugVcpuError}};
+    use crate::api_server::parsed_request::tests::vmm_action_from_request;
+
+    use super::super::*;
+
+    #[test]
+    fn test_parse_put_hotplug() {
+        // Case 1. Invalid resource type
+        parse_put_hotplug(&Body::new("invalid body"), Some("invalid")).unwrap_err();
+
+        // Case 2. No resource type
+        parse_put_hotplug(&Body::new("invalid body"), None).unwrap_err();
```

Comment:
nit: Since these are explicitly testing incorrect values for the "resource type" parameter, we should use valid bodies, to ensure that the error we are receiving is actually related to the resource type (and not some other invalid configuration we happen to also pass). Alternatively, you could do some assertion on the returned error (e.g. by checking the error variant is, say, `InvalidPathMethod` in the first case, and looking at the method/error string)